tests/agent/actions/checkpoints/test_run_checkpoint_action.py

```diff
@@ -97,3 +105,23 @@ def test_run_checkpoint_action_with_splitter_options_returns_action_result(
             type="SuiteValidationResult",
         ),
     ]
+
+
+def test_run_checkpoint_action_raises_on_test_connection_failure(
+    context, checkpoint_id, datasource_names
+):
+    mock_datasource = MagicMock()
```

Comment:
```suggestion
    mock_datasource = MagicMock(spec=Datasource)
```
MagicMock without a spec is a dangerous footgun.
It will recursively create mocks for any attribute that you access.